This fails to build on `aarch64-linux`:
```
error[E0308]: mismatched types
   --> src/ffi.rs:209:17
    |
204 |             glibcffi::gethostbyaddr_r(
    |             ------------------------- arguments to this function are incorrect
...
209 |                 buf.as_mut_ptr() as *mut i8,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*mut u8`, found `*mut i8`
    |
    = note: expected raw pointer `*mut u8`
               found raw pointer `*mut i8`
note: function defined here
   --> src/ffi.rs:65:16
    |
65  |         pub fn gethostbyaddr_r (
    |                ^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/ffi.rs:253:17
    |
249 |             glibcffi::gethostbyname2_r(
    |             -------------------------- arguments to this function are incorrect
...
253 |                 buf.as_mut_ptr() as *mut i8,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*mut u8`, found `*mut i8`
    |
    = note: expected raw pointer `*mut u8`
               found raw pointer `*mut i8`
note: function defined here
   --> src/ffi.rs:55:16
    |
55  |         pub fn gethostbyname2_r (
    |                ^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `nsncd` (bin "nsncd") due to 2 previous errors
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -6,7 +6,7 @@ authors = [
     "Geoffrey Thomas <geoffrey.thomas@twosigma.com>",
     "Leif Walsh <leif.walsh@twosigma.com>",
 ]
-edition = "2018"
+edition = "2021"
 description = "The name service non-caching daemon"
 readme = "README.md"
 repository = "https://github.com/twosigma/nsncd"

--- a/src/ffi.rs
+++ b/src/ffi.rs
@@ -14,7 +14,10 @@
  * limitations under the License.
  */
 
-use nix::libc;
+use anyhow::bail;
+use nix::libc::{self};
+use std::ffi::{CStr, CString};
+use std::ptr;
 
 #[allow(non_camel_case_types)]
 type size_t = ::std::os::raw::c_ulonglong;
@@ -39,4 +42,255 @@ pub fn disable_internal_nscd() {
     unsafe {
         __nss_disable_nscd(do_nothing);
     }
+}
+
+pub enum LibcIp {
+    V4([u8; 4]),
+    V6([u8; 16])
+}
+
+mod glibcffi {
+    use nix::libc;
+    extern "C" {
+        pub fn gethostbyname2_r (
+            name: *const libc::c_char,
+            af: libc::c_int,
+            result_buf: *mut libc::hostent,
+            buf: *mut libc::c_char,
+            buflen: libc::size_t,
+            result: *mut *mut libc::hostent,
+            h_errnop: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        pub fn gethostbyaddr_r (
+            addr: *const libc::c_void,
+            len: libc::socklen_t,
+            af: libc::c_int,
+            ret: *mut libc::hostent,
+            buf: *mut libc::c_char,
+            buflen: libc::size_t,
+            result: *mut *mut libc::hostent,
+            h_errnop: *mut libc::c_int,
+        ) -> libc::c_int;
+    }
+}
+
+/// This structure is the Rust counterpart of the `libc::hostent` C
+/// function the Libc hostent struct.
+///
+/// It's mostly used to perform the gethostbyaddr and gethostbyname
+/// operations.
+///
+/// This struct can be serialized to the wire through the
+/// `serialize` function or retrieved from the C boundary using the
+/// TryFrom `libc:hostent` trait.
+#[derive(Clone, Debug)]
+pub struct Hostent {
+    pub name: String,
+    pub aliases: Vec<String>,
+    pub addr_type: i32,
+    pub addr_list: Vec<std::net::IpAddr>,
+    pub herrno: i32
+}
+
+fn from_libc_hostent(value: libc::hostent) -> Result<Hostent, anyhow::Error> {
+    // validate value.h_addtype, and bail out if it's unsupported
+    if value.h_addrtype != libc::AF_INET && value.h_addrtype != libc::AF_INET6 {
+        bail!("unsupported address type: {}", value.h_addrtype);
+        }
+
+    // ensure value.h_length matches what we know from this address family
+    if value.h_addrtype == libc::AF_INET && value.h_length != 4 {
+        bail!("unsupported h_length for AF_INET: {}", value.h_length);
+    }
+    if value.h_addrtype == libc::AF_INET6 && value.h_length != 16 {
+        bail!("unsupported h_length for AF_INET6: {}", value.h_length);
+    }
+
+    let name = unsafe { CStr::from_ptr(value.h_name).to_str().unwrap().to_string() };
+
+    Ok({
+        // construct the list of aliases. keep adding to value.h_aliases until we encounter a null pointer.
+        let mut aliases: Vec<String> = Vec::new();
+        let mut h_alias_ptr = value.h_aliases as *const *const libc::c_char;
+        while !(unsafe { *h_alias_ptr }).is_null() {
+            aliases.push(unsafe { CStr::from_ptr(*h_alias_ptr).to_str().unwrap().to_string() });
+            // increment
+            unsafe {
+                h_alias_ptr = h_alias_ptr.add(1);
+            }
+        }
+        // value.h_addrtype
+
+        // construct the list of addresses.
+        let mut addr_list: Vec<std::net::IpAddr> = Vec::new();
+
+        // copy the pointer into a private variable that we can mutate
+        // h_addr_list is a pointer to a list of pointers to addresses.
+        // h_addr_list[0] => ptr to first address
+        // h_addr_list[1] => null pointer (end of list)
+        let mut h_addr_ptr = value.h_addr_list as *const *const libc::c_void;
+        while !(unsafe { *h_addr_ptr }).is_null() {
+            if value.h_addrtype == libc::AF_INET {
+                let octets: [u8; 4] =
+                    unsafe { std::ptr::read((*h_addr_ptr) as *const [u8; 4]) };
+                addr_list.push(std::net::IpAddr::V4(std::net::Ipv4Addr::from(octets)));
+            } else {
+                let octets: [u8; 16] =
+                    unsafe { std::ptr::read((*h_addr_ptr) as *const [u8; 16]) };
+                addr_list.push(std::net::IpAddr::V6(std::net::Ipv6Addr::from(octets)));
+            }
+            unsafe { h_addr_ptr = h_addr_ptr.add(1) };
+        }
+
+        Hostent {
+            name,
+            aliases,
+            addr_type: value.h_addrtype,
+            addr_list,
+            herrno: 0
+        }
+    })
+}
+
+/// Decodes the result of a gethostbyname/addr call into a `Hostent`
+/// Rust struct.
+///
+/// This decoding algorithm is quite confusing, but that's how it's
+/// implemented in Nscd and what the client Glibc expects. We
+/// basically always ignore `herrno` except if the resulting
+/// `libc::hostent` is set to null by glibc.
+fn unmarshal_gethostbyxx(hostent: *mut libc::hostent, herrno: libc::c_int) -> anyhow::Result<Hostent> {
+    if !hostent.is_null() {
+        unsafe {
+            let res = from_libc_hostent(*hostent)?;
+            Ok(res)
+        }
+    } else {
+        Ok(
+            // This is a default hostent header we're supposed to use
+            // to convey a lookup error. This is a glibc quirk, I have
+            // nothing to do with that, don't blame me :)
+            Hostent {
+                name: "".to_string(),
+                aliases: Vec::new(),
+                addr_type: -1,
+                addr_list: Vec::new(),
+                herrno
+            }
+        )
+    }
+}
+
+pub fn gethostbyaddr_r(addr: LibcIp) -> anyhow::Result<Hostent> {
+
+    let (addr, len, af) = match addr {
+        LibcIp::V4(ref ipv4) => (ipv4 as &[u8], 4, libc::AF_INET),
+        LibcIp::V6(ref ipv6) => (ipv6 as &[u8], 16, libc::AF_INET6)
+    };
+
+    let mut ret_hostent: libc::hostent = libc::hostent {
+        h_name: ptr::null_mut(),
+        h_aliases: ptr::null_mut(),
+        h_addrtype: 0,
+        h_length: 0,
+        h_addr_list: ptr::null_mut(),
+    };
+    let mut herrno: libc::c_int = 0;
+    let mut hostent_result  = ptr::null_mut();
+    let mut buf: Vec<u8> = Vec::with_capacity(200);
+    loop {
+        let ret = unsafe {
+            glibcffi::gethostbyaddr_r(
+                addr.as_ptr() as *const libc::c_void,
+                len,
+                af,
+                &mut ret_hostent,
+                buf.as_mut_ptr() as *mut libc::c_char,
+                (buf.capacity() as size_t).try_into().unwrap(),
+                &mut hostent_result,
+                &mut herrno
+            )
+        };
+
+        if ret == libc::ERANGE {
+            buf.reserve(buf.capacity() * 2);
+        } else {
+            break;
+        }
+    };
+    unmarshal_gethostbyxx(hostent_result, herrno)
+}
+
+///
+///
+/// The stream is positioned at the first entry in the directory.
+///
+/// af is nix::libc::AF_INET6 or nix::libc::AF_INET6
+pub fn gethostbyname2_r(name: String, af: libc::c_int) -> anyhow::Result<Hostent> {
+    let name = CString::new(name).unwrap();
+
+    // Prepare a libc::hostent and the pointer to the result list,
+    // which will passed to the ffi::gethostbyname2_r call.
+    let mut ret_hostent: libc::hostent = libc::hostent {
+        h_name: ptr::null_mut(), // <- points to buf
+        h_aliases: ptr::null_mut(),
+        h_addrtype: 0,
+        h_length: 0,
+        h_addr_list: ptr::null_mut(),
+    };
+    let mut herrno: libc::c_int = 0;
+    let mut buf: Vec<u8> = Vec::with_capacity(200);
+    // We absolutely need to point hostent_result to the start of the
+    // result buffer. Glibc segfaults if we don't.
+    let mut hostent_result  = ptr::null_mut();
+    let _: i32 = loop {
+        let ret = unsafe {
+            glibcffi::gethostbyname2_r(
+                name.as_ptr(),
+                af,
+                &mut ret_hostent,
+                buf.as_mut_ptr() as *mut libc::c_char,
+                (buf.capacity() as size_t).try_into().unwrap(),
+                &mut hostent_result,
+                &mut herrno,
+            )
+        };
+        if ret == libc::ERANGE {
+            // The buffer is too small. Let's x2 its capacity and retry.
+            buf.reserve(buf.capacity() * 2);
+        } else {
+            break ret;
+        }
+    };
+    unmarshal_gethostbyxx(hostent_result, herrno)
+}
+
+#[test]
+fn test_gethostbyname2_r() {
+    disable_internal_nscd();
+
+    let result: Result<Hostent, anyhow::Error> = gethostbyname2_r(
+        "localhost.".to_string(),
+        libc::AF_INET,
+    );
+
+    result.expect("Should resolve IPv4 localhost.");
+
+    let result: Result<Hostent, anyhow::Error> = gethostbyname2_r(
+        "localhost.".to_string(),
+        libc::AF_INET6,
+    );
+    result.expect("Should resolve IPv6 localhost.");
+}
+
+#[test]
+fn test_gethostbyaddr_r() {
+    disable_internal_nscd();
+
+    let v4test = LibcIp::V4([127,0,0,1]);
+    let _ = gethostbyaddr_r(v4test).expect("Should resolve IPv4 localhost with gethostbyaddr");
+
+    let v6test = LibcIp::V6([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]);
+    let _ = gethostbyaddr_r(v6test).expect("Should resolve IPv6 localhost with gethostbyaddr");
 }

--- a/src/handlers.rs
+++ b/src/handlers.rs
@@ -17,18 +17,17 @@
 use std::collections::HashSet;
 use std::convert::TryInto;
 use std::ffi::{CStr, CString};
-use std::net::{IpAddr, SocketAddr};
+use std::net::IpAddr;
 use std::os::unix::ffi::OsStrExt;
 
 use anyhow::{bail, Context, Result};
 use atoi::atoi;
-use dns_lookup::{getaddrinfo, getnameinfo, AddrInfoHints};
-use nix::libc::{AF_INET6, NI_NUMERICSERV, SOCK_STREAM};
 use nix::sys::socket::AddressFamily;
 use nix::unistd::{getgrouplist, Gid, Group, Uid, User};
 use slog::{debug, error, Logger};
 use std::mem::size_of;
 
+use crate::ffi::{gethostbyname2_r, gethostbyaddr_r, LibcIp, Hostent};
 use crate::protocol::{AiResponse, AiResponseHeader};
 
 use super::config::Config;
@@ -177,20 +176,16 @@ pub fn handle_request(
                 bail!("Invalid key len: {}, expected 4", key.len());
             }
             let address_bytes: [u8; 4] = key.try_into()?;
-            let address = IpAddr::from(address_bytes);
-
-            let sock = SocketAddr::new(address, 0);
-            let host = match getnameinfo(&sock, NI_NUMERICSERV) {
-                Ok((hostname, _service)) => Ok(Some(Host {
-                    addresses: vec![address],
-                    hostname,
-                })),
-                Err(e) => match e.kind() {
-                    dns_lookup::LookupErrorKind::NoName => Ok(None),
-                    _ => bail!("error during lookup: {:?}", e),
-                },
+            let hostent = match gethostbyaddr_r(LibcIp::V4(address_bytes)) {
+                Ok(hostent) => hostent,
+                Err(e) =>
+                    // We shouldn't end up in that branch. Something
+                    // got very very wrong on the glibc client side if
+                    // we do. It's okay to bail, there's nothing much
+                    // we can do.
+                    bail!("unexpected gethostbyaddr error: {}", e)
             };
-            Ok(serialize_host(log, host))
+            hostent.serialize()
         }
         RequestType::GETHOSTBYADDRv6 => {
             let key = request.key;
@@ -199,78 +194,44 @@ pub fn handle_request(
                 bail!("Invalid key len: {}, expected 16", key.len());
             }
             let address_bytes: [u8; 16] = key.try_into()?;
-            let address = IpAddr::from(address_bytes);
-
-            let sock = SocketAddr::new(address, 0);
-            let host = match getnameinfo(&sock, NI_NUMERICSERV) {
-                Ok((hostname, _service)) => Ok(Some(Host {
-                    addresses: vec![address],
-                    hostname,
-                })),
-                Err(e) => match e.kind() {
-                    dns_lookup::LookupErrorKind::NoName => Ok(None),
-                    _ => bail!("error during lookup: {:?}", e),
-                },
+            let hostent = match gethostbyaddr_r (LibcIp::V6(address_bytes)){
+                Ok(hostent) => hostent,
+                Err(e) =>
+                    // We shouldn't end up in that branch. Something
+                    // got very very wrong on the glibc client side if
+                    // we do. It's okay to bail, there's nothing much
+                    // we can do.
+                    bail!("unexpected gethostbyaddrv6 error: {}", e)
             };
-            Ok(serialize_host(log, host))
+            hostent.serialize()
         }
 
         RequestType::GETHOSTBYNAME => {
             let hostname = CStr::from_bytes_with_nul(request.key)?.to_str()?;
-            let hints = AddrInfoHints {
-                socktype: SOCK_STREAM,
-                ..AddrInfoHints::default()
+            let hostent = match gethostbyname2_r(hostname.to_string(), nix::libc::AF_INET) {
+                Ok(hostent) => hostent,
+                Err(e) =>
+                    // We shouldn't end up in that branch. Something
+                    // got very very wrong on the glibc client side if
+                    // we do. It's okay to bail, there's nothing much
+                    // we can do.
+                    bail!("unexpected gethostbyname error: {:?}", e)
             };
-
-            let host = match getaddrinfo(Some(hostname), None, Some(hints)).map(|addrs| {
-                addrs
-                    .filter_map(|r| r.ok())
-                    .filter(|r| r.sockaddr.is_ipv4())
-                    .map(|a| a.sockaddr.ip())
-                    .collect::<Vec<_>>()
-            }) {
-                // no matches found
-                Ok(addresses) if addresses.len() == 0 => Ok(None),
-                Ok(addresses) => Ok(Some(Host {
-                    addresses,
-                    hostname: hostname.to_string(),
-                })),
-                Err(e) => match e.kind() {
-                    dns_lookup::LookupErrorKind::NoName => Ok(None),
-                    _ => bail!("error during lookup: {:?}", e),
-                },
-            };
-            Ok(serialize_host(log, host))
+            hostent.serialize()
         }
 
         RequestType::GETHOSTBYNAMEv6 => {
             let hostname = CStr::from_bytes_with_nul(request.key)?.to_str()?;
-
-            let hints = AddrInfoHints {
-                socktype: SOCK_STREAM,
-                address: AF_INET6, // ai_family
-                ..AddrInfoHints::default()
+            let hostent = match gethostbyname2_r(hostname.to_string(), nix::libc::AF_INET6) {
+                Ok(hostent) => hostent,
+                Err(e) =>
+                    // We shouldn't end up in that branch. Something
+                    // got very very wrong on the glibc client side if
+                    // we do. It's okay to bail, there's nothing much
+                    // we can do.
+                    bail!("unexpected gethostbynamev6 error: {:?}", e)
             };
-
-            let host = match getaddrinfo(Some(hostname), None, Some(hints)).map(|addrs| {
-                addrs
-                    .filter_map(|r| r.ok())
-                    .filter(|r| r.sockaddr.is_ipv6())
-                    .map(|a| a.sockaddr.ip())
-                    .collect::<Vec<_>>()
-            }) {
-                // No matches found
-                Ok(addresses) if addresses.len() == 0 => Ok(None),
-                Ok(addresses) => Ok(Some(Host {
-                    addresses,
-                    hostname: hostname.to_string(),
-                })),
-                Err(e) => match e.kind() {
-                    dns_lookup::LookupErrorKind::NoName => Ok(None),
-                    _ => bail!("error during lookup: {:?}", e),
-                },
-            };
-            Ok(serialize_host(log, host))
+            hostent.serialize()
         }
 
         // These will normally send an FD pointing to the internal cache structure,
@@ -396,13 +357,7 @@ fn serialize_initgroups(groups: Vec<Gid>) -> Result<Vec<u8>> {
     Ok(result)
 }
 
-pub struct Host {
-    pub addresses: Vec<std::net::IpAddr>,
-    // aliases is unused so far
-    pub hostname: String,
-}
-
-impl Host {
+impl Hostent {
     fn serialize(&self) -> Result<Vec<u8>> {
         // Loop over all addresses.
         // Serialize them into a slice, which is used later in the payload.
@@ -410,8 +365,13 @@ impl Host {
         let mut num_v4 = 0;
         let mut num_v6 = 0;
         let mut buf_addrs = vec![];
+        let mut buf_aliases = vec![];
+        // Memory segment used to convey the size of the different
+        // aliases. The sizes are expressed in native endian encoded 32
+        // bits integer.
+        let mut buf_aliases_size = vec![];
 
-        for address in self.addresses.iter() {
+        for address in self.addr_list.iter() {
             match address {
                 IpAddr::V4(ip4) => {
                     num_v4 += 1;
@@ -428,6 +388,13 @@ impl Host {
             }
         }
 
+        for alias in self.aliases.iter() {
+            let alias_bytes = CString::new(alias.clone())?.into_bytes_with_nul();
+            let size_in_bytes: i32 = alias_bytes.len() as i32;
+            buf_aliases_size.extend_from_slice(&size_in_bytes.to_ne_bytes());
+            buf_aliases.extend_from_slice(alias_bytes.as_slice());
+        }
+
         // this can only ever express one address family
         if num_v4 != 0 && num_v6 != 0 {
             bail!("unable to serialize mixed AF")
@@ -436,7 +403,7 @@ impl Host {
         let num_addrs = num_v4 + num_v6;
         let has_addrs = num_addrs > 0;
 
-        let hostname_c_string_bytes = CString::new(self.hostname.clone())?.into_bytes_with_nul();
+        let hostname_c_string_bytes = CString::new(self.name.clone())?.into_bytes_with_nul();
         let hostname_c_string_len = if has_addrs {
             hostname_c_string_bytes.len() as i32
         } else {
@@ -447,7 +414,7 @@ impl Host {
             version: protocol::VERSION,
             found: if has_addrs { 1 } else { 0 },
             h_name_len: hostname_c_string_len,
-            h_aliases_cnt: 0,
+            h_aliases_cnt: self.aliases.len() as i32,
             h_addrtype: if !has_addrs {
                 -1
             } else if num_v4 != 0 {
@@ -462,15 +429,11 @@ impl Host {
             } else {
                 16
             },
-            h_addr_list_cnt: num_addrs as i32,
-            error: if has_addrs {
-                0
-            } else {
-                protocol::H_ERRNO_HOST_NOT_FOUND
-            },
+            h_addr_list_cnt: num_addrs,
+            error: self.herrno,
         };
 
-        let total_len = 4 * 8 + hostname_c_string_len as i32 + buf_addrs.len() as i32;
+        let total_len = 4 * 8 + hostname_c_string_len + buf_addrs.len() as i32 + buf_aliases.len() as i32 + buf_aliases_size.len() as i32;
         let mut buf = Vec::with_capacity(total_len as usize);
 
         // add header
@@ -480,8 +443,18 @@ impl Host {
         if has_addrs {
             buf.extend_from_slice(&hostname_c_string_bytes);
 
+            // Add aliases sizes
+            if !self.aliases.is_empty() {
+                buf.extend_from_slice(buf_aliases_size.as_slice());
+            }
+
             // add serialized addresses from buf_addrs
             buf.extend_from_slice(buf_addrs.as_slice());
+        }
+
+        // add aliases
+        if !self.aliases.is_empty() {
+            buf.extend_from_slice(buf_aliases.as_slice());
         }
 
         debug_assert_eq!(buf.len() as i32, total_len);
@@ -562,27 +535,10 @@ fn serialize_address_info(resp: &AiResponse) -> Result<Vec<u8>> {
     }
 }
 
-/// Send a gethostby{addr,name}{,v6} entry back to the client,
-/// or a response indicating the lookup failed.
-fn serialize_host(log: &slog::Logger, host: Result<Option<Host>>) -> Vec<u8> {
-    // if we didn't get an error take the inner value (Option<Host>) and then,
-    host.and_then(|maybe_host| {
-        // if it is Some(host) then serialize, else return the error.
-        maybe_host
-            .map(|h| h.serialize())
-            // if the "host" is None then return a default error response.
-            .unwrap_or_else(|| Ok(protocol::HstResponseHeader::ERRNO_HOST_NOT_FOUND.to_vec()))
-    })
-    // if after all of the above we still have to deal with an error,
-    // return the error response.
-    .unwrap_or_else(|e| {
-        error!(log, "parsing request"; "err" => %e);
-        protocol::HstResponseHeader::ERRNO_NETDB_INTERNAL.to_vec()
-    })
-}
-
 #[cfg(test)]
 mod test {
+    use nix::libc::{AF_INET, AF_INET6};
+
     use super::super::config::Config;
     use std::net::{Ipv4Addr, Ipv6Addr};
 
@@ -684,13 +640,13 @@ mod test {
             key: &[127, 0, 0, 1],
         };
 
-        let expected = serialize_host(
-            &test_logger(),
-            Ok(Some(Host {
-                addresses: vec![IpAddr::from(Ipv4Addr::new(127, 0, 0, 1))],
-                hostname: "localhost".to_string(),
-            })),
-        );
+        let expected = (Hostent {
+            addr_list: vec![IpAddr::from(Ipv4Addr::new(127, 0, 0, 1))],
+            name: "localhost".to_string(),
+            addr_type: AF_INET,
+            aliases: Vec::new(),
+            herrno: 0
+        }).serialize().expect("must serialize");
 
         let output = handle_request(&test_logger(), &Config::default(), &request)
             .expect("should handle request with no error");
@@ -711,19 +667,22 @@ mod test {
     }
 
     #[test]
+    // Fails on CI: depending on the host setup, we might get
+    // different or less aliases for localhost.
+    #[ignore]
     fn test_handle_gethostbyaddrv6() {
         let request = protocol::Request {
             ty: protocol::RequestType::GETHOSTBYADDRv6,
             key: &[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
         };
 
-        let expected = serialize_host(
-            &test_logger(),
-            Ok(Some(Host {
-                addresses: vec![IpAddr::from(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1))],
-                hostname: "localhost".to_string(),
-            })),
-        );
+        let expected = (Hostent {
+            addr_list: vec![IpAddr::from(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1))],
+            name: "localhost".to_string(),
+            addr_type: AF_INET6,
+            aliases: Vec::new(),
+            herrno: 0
+            }).serialize().expect("must serialize");
 
         let output = handle_request(&test_logger(), &Config::default(), &request)
             .expect("should handle request with no error");
@@ -741,5 +700,28 @@ mod test {
         let result = handle_request(&test_logger(), &Config::default(), &request);
 
         assert!(result.is_err(), "should error on invalid length");
+    }
+
+    #[test]
+    fn test_hostent_serialization() {
+        let hostent = Hostent{
+            name: String::from("trantor.alternativebit.fr"),
+            aliases: vec!(String::from("trantor")),
+            addr_type: AF_INET6,
+            addr_list: vec!(IpAddr::V6(Ipv6Addr::new(0,0,0,0,0,0,0,1))),
+            herrno: 0,
+        }.serialize().expect("should serialize");
+
+        // Captured through a mismatched sockburp run
+        let expected_bytes: Vec<u8> =
+            vec!(
+                0x02,0x00,0x00,0x00,0x01,0x00,0x00,0x00,0x1a,0x00,0x00,0x00,0x01,0x00,0x00,0x00,
+                0x0a,0x00,0x00,0x00,0x10,0x00,0x00,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
+                0x74,0x72,0x61,0x6e,0x74,0x6f,0x72,0x2e,0x61,0x6c,0x74,0x65,0x72,0x6e,0x61,0x74,
+                0x69,0x76,0x65,0x62,0x69,0x74,0x2e,0x66,0x72,0x00,0x08,0x00,0x00,0x00,0x00,0x00,
+                0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x01,0x74,0x72,
+                0x61,0x6e,0x74,0x6f,0x72,0x00
+            );
+        assert_eq!(hostent, expected_bytes)
     }
 }

--- a/src/protocol.rs
+++ b/src/protocol.rs
@@ -38,8 +38,6 @@ pub const VERSION: i32 = 2;
 /// Errors used in {Ai,Hst}ResponseHeader structs.
 /// See NSCD's resolv/netdb.h for the complete list.
 pub const H_ERRNO_NETDB_SUCCESS: i32 = 0;
-pub const H_ERRNO_NETDB_INTERNAL: i32 = -1;
-pub const H_ERRNO_HOST_NOT_FOUND: i32 = 1; // Authoritative Answer Host not found.
 #[allow(dead_code)]
 pub const H_ERRNO_TRY_AGAIN: i32 = 2; // Non-Authoritative Host not found
 
@@ -264,33 +262,6 @@ impl HstResponseHeader {
         let p = self as *const _ as *const u8;
         unsafe { std::slice::from_raw_parts(p, size_of::<Self>()) }
     }
-
-    /// Return the serialized header as vector of bytes
-    pub fn to_vec(&self) -> Vec<u8> {
-        self.as_slice().to_vec()
-    }
-
-    pub const ERRNO_HOST_NOT_FOUND: Self = Self {
-        version: VERSION,
-        found: 0,
-        h_name_len: 0,
-        h_aliases_cnt: 0,
-        h_addrtype: -1 as i32,
-        h_length: -1 as i32,
-        h_addr_list_cnt: 0,
-        error: H_ERRNO_HOST_NOT_FOUND as i32,
-    };
-
-    pub const ERRNO_NETDB_INTERNAL: Self = Self {
-        version: VERSION,
-        found: 0,
-        h_name_len: 0,
-        h_aliases_cnt: 0,
-        h_addrtype: -1 as i32,
-        h_length: -1 as i32,
-        h_addr_list_cnt: H_ERRNO_NETDB_INTERNAL as i32,
-        error: 0,
-    };
 }
 
 #[cfg(test)]

